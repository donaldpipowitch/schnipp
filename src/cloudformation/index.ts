import {
  CloudFormationCustomResourceEvent,
  CloudFormationCustomResourceResponse
} from 'aws-lambda';
import { IHandler } from '../';

export type CloudFormationCustomResourceHandler = IHandler<
  CloudFormationCustomResourceEvent,
  CloudFormationCustomResourceResponse
>;
