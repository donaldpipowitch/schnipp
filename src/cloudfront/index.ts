import {
  CloudFrontRequestEvent,
  CloudFrontRequestResult,
  CloudFrontResponseEvent,
  CloudFrontResponseResult
} from 'aws-lambda';
import { IHandler } from '..';

export type CloudFrontRequestHandler = IHandler<
  CloudFrontRequestEvent,
  CloudFrontRequestResult
>;

export type CloudFrontResponseHandler = IHandler<
  CloudFrontResponseEvent,
  CloudFrontResponseResult
>;
