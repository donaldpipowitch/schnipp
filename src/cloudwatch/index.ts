import { CloudWatchLogsEvent } from 'aws-lambda';
import { IHandler } from '../';

export type CloudwatchHandler = IHandler<CloudWatchLogsEvent, void>;
