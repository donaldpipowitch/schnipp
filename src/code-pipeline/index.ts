import { IHandler } from '..';
import { CodePipelineEvent } from 'aws-lambda';

export type CodePipelineHandler = IHandler<CodePipelineEvent>;
