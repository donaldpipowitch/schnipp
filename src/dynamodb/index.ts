import { DynamoDBStreamEvent } from 'aws-lambda';
import { IHandler } from '../';

export type DynamoDBStreamHandler = IHandler<DynamoDBStreamEvent, void>;
