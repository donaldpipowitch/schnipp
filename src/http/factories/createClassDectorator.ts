import { DecoratedExceptionFilter, Guard, MetaData } from '..';
import { Newable } from 'simple-ts-di';

// export const createClassDecorator = (
//   decorator: (data: Partial<ControllerData>) => void
// ): ClassDecorator => (target: any) => {
//   decorator(getControllerMetaData(target));
// };

export const CONTROLLER = Symbol('controller');

export type ControllerData = {
  basePath: string;
  headers: { [key: string]: string };
  filters: Array<DecoratedExceptionFilter>;
  guards: Array<Newable<Guard>>;
  metaData: MetaData;
};

export const getControllerMetaData = (controller: any) => {
  let data: ControllerData | null = Reflect.getMetadata(CONTROLLER, controller);

  if (data == null) {
    data = {
      basePath: '/',
      filters: [],
      guards: [],
      metaData: {},
      headers: {}
    };
    Reflect.defineMetadata(CONTROLLER, data, controller);
  }

  return data;
};
