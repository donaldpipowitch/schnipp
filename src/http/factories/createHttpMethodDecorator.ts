import * as urlJoin from 'url-join';
import { APIGatewayEvent, Context, APIGatewayProxyResult } from 'aws-lambda';
import { Newable } from 'simple-ts-di';
import { HttpVerb } from '../utils/AuthPolicy';

const METHODS = Symbol('methods');

export type ParamInjector<T> = {
  (e: APIGatewayEvent, metaData: MetaData, context: Context): T;
};

export type InjectedParam<T> = {
  index: number;
  injector: ParamInjector<T>;
};

export interface ExceptionFilter {
  catch(
    exception: any,
    event: APIGatewayEvent,
    metaData: MetaData,
    context: Context
  ): Promise<APIGatewayProxyResult> | APIGatewayProxyResult;
}

export type DecoratedExceptionFilter = {
  exceptions: Array<any>;
  filter: Newable<ExceptionFilter>;
};

export interface Guard {
  canActivate(
    event: APIGatewayEvent,
    metaData: MetaData,
    context: Context
  ): Promise<boolean> | boolean;
}

export type MetaData = { [key: string]: any };

export type MethodMetaData = {
  statusCode?: number;
  method?: HttpVerb;
  path: string;
  propertyKey: string;
  headers: { [key: string]: string };
  paramaters: Array<InjectedParam<any>>;
  filters: Array<DecoratedExceptionFilter>;
  guards: Array<Newable<Guard>>;
  metaData: MetaData;
};

export const createHttpMethodDecorator = (method: HttpVerb) => (
  path: string = '/'
): MethodDecorator => <T>(
  target: any,
  propertyKey: string | symbol,
  descriptor: TypedPropertyDescriptor<T>
) => {
  const methodMetaData = getMethodMetaData(propertyKey.toString(), target);
  if (methodMetaData.statusCode == null) {
    methodMetaData.statusCode = method === 'POST' ? 201 : 200;
  }
  methodMetaData.method = method;
  methodMetaData.path = path;
  return descriptor;
};

export const getMethodMetaData = (propertyKey: string, controller: any) => {
  let methods: Array<MethodMetaData> | null = Reflect.getMetadata(
    METHODS,
    controller
  );

  if (methods == null) {
    methods = [];
    Reflect.defineMetadata(METHODS, methods, controller);
  }

  let data = methods.find(x => x.propertyKey === propertyKey);

  if (data == null) {
    data = {
      path: '/',
      propertyKey,
      filters: [],
      paramaters: [],
      guards: [],
      headers: {},
      metaData: {}
    };
    methods.push(data);
  }

  return data;
};

export const findMethod = (
  basePath: string,
  resourcePath: string,
  method: string,
  controller: any
) => {
  const m: Array<MethodMetaData> = Reflect.getMetadata(METHODS, controller);

  const res = m.find(
    x =>
      x.method === method &&
      urlJoin(basePath, x.path!.replace(/\/$/, '')) === resourcePath
  );

  if (res == null) {
    return m.find(x => x.method === '*');
  }

  return res;
};
