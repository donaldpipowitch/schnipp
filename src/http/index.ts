import {
  APIGatewayEvent,
  APIGatewayProxyResult,
  CustomAuthorizerEvent,
  CustomAuthorizerResult
} from 'aws-lambda';
import { IHandler } from '..';

export type APIGatewayHandler = IHandler<
  APIGatewayEvent,
  APIGatewayProxyResult
>;

export type AuthorizeHandler = IHandler<
  CustomAuthorizerEvent,
  CustomAuthorizerResult
>;

export { RestController } from './RestController';

export {
  InjectedParam,
  MetaData,
  MethodMetaData,
  ParamInjector,
  DecoratedExceptionFilter,
  Guard,
  ExceptionFilter
} from './factories/createHttpMethodDecorator';

export { Catch } from './decorators/catch';
export {
  GET,
  POST,
  DELETE,
  PUT,
  PATCH,
  OPTIONS,
  HEAD
} from './decorators/methods';
export { Path } from './decorators/path';
export { Param } from './decorators/param';
export { Query } from './decorators/query';
export { Header } from './decorators/header';
export { StatusCode } from './decorators/statusCode';
export { AuthContext } from './decorators/authContext';
export { Meta } from './decorators/meta';
export { Body } from './decorators/body';
export { Use } from './decorators/use';

export { HttpException } from './exceptions/HttpException';
export { NotFoundException } from './exceptions/NotFoundException';
export { ForbiddenException } from './exceptions/ForbiddenException';

export { AuthPolicy, HttpVerb, Effect } from './utils/AuthPolicy';
export { createResponse } from './utils/createResponse';
