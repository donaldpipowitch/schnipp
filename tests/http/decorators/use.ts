import 'reflect-metadata';
import { Use, createResponse, Guard, ExceptionFilter } from '../../../src/http';
import { getMethodMetaData } from '../../../src/http/factories/createHttpMethodDecorator';
import { getControllerMetaData } from '../../../src/http/factories/createClassDectorator';

describe('@Use', () => {
  it('should add header metadata for methods and classes', () => {
    class TestGuard implements Guard {
      canActivate() {
        return true;
      }
    }

    class Filter implements ExceptionFilter {
      catch() {
        return createResponse(200);
      }
    }

    @Use(TestGuard, Filter)
    class Test {
      @Use(TestGuard, Filter)
      test() {}
    }

    const f = new Test();

    expect(getMethodMetaData('test', f)).toMatchSnapshot();
    expect(getControllerMetaData(Test)).toMatchSnapshot();
  });

  it('should throw on wrong usage', () => {
    expect(() => Use()()).toThrow();
    expect(() => Use()(1, 2, 3, 4)).toThrow();
  });
});
