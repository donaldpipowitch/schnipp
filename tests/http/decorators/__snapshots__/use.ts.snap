// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@Use should add header metadata for methods and classes 1`] = `
Object {
  "filters": Array [
    Object {
      "exceptions": Array [],
      "filter": [Function],
    },
  ],
  "guards": Array [
    [Function],
  ],
  "headers": Object {},
  "metaData": Object {},
  "paramaters": Array [],
  "path": "/",
  "propertyKey": "test",
}
`;

exports[`@Use should add header metadata for methods and classes 2`] = `
Object {
  "basePath": "/",
  "filters": Array [
    Object {
      "exceptions": Array [],
      "filter": [Function],
    },
  ],
  "guards": Array [
    [Function],
  ],
  "headers": Object {},
  "metaData": Object {},
}
`;
